src/Domain/FunctionsOfTime/FunctionOfTimeHelpers.cpp

```diff
@@ -0,0 +1,143 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Domain/FunctionsOfTime/FunctionOfTimeHelpers.hpp"
+
+#include <pup.h>
+#include <pup_stl.h>
+
+#include "Utilities/EqualWithinRoundoff.hpp"
+#include "Utilities/ErrorHandling/Error.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+namespace domain::FunctionsOfTime::FunctionOfTimeHelpers {
+template <size_t MaxDerivPlusOne, bool StoreCoefs>
+StoredInfo<MaxDerivPlusOne, StoreCoefs>::StoredInfo(
+    const double t,
+    std::array<DataVector, MaxDerivPlusOne> init_quantities) noexcept
+    : time(t), stored_quantities{std::move(init_quantities)} {
+  // Convert derivs to coefficients for polynomial evaluation
+  // The coefficient of x^N is the Nth deriv rescaled by 1/factorial(N)
+  double fact = 1.0;
+  if (MaxDerivPlusOne > 2) {
+    for (size_t j = 2; j < MaxDerivPlusOne; j++) {
+      fact *= j;
+      gsl::at(stored_quantities, j) /= fact;
+    }
+  }
+}
+
+template <size_t MaxDerivPlusOne>
+StoredInfo<MaxDerivPlusOne, false>::StoredInfo(
+    const double t,
+    std::array<DataVector, MaxDerivPlusOne> init_quantities) noexcept
+    : time(t), stored_quantities{std::move(init_quantities)} {}
+
+template <size_t MaxDerivPlusOne, bool StoreCoefs>
+void StoredInfo<MaxDerivPlusOne, StoreCoefs>::pup(PUP::er& p) noexcept {
+  p | time;
+  p | stored_quantities;
+}
+
+template <size_t MaxDerivPlusOne>
+void StoredInfo<MaxDerivPlusOne, false>::pup(PUP::er& p) noexcept {
+  p | time;
+  p | stored_quantities;
+}
+
+void reset_expiration_time(const gsl::not_null<double*> prev_expiration_time,
+                           const double next_expiration_time) noexcept {
+  if (next_expiration_time < *prev_expiration_time) {
+    ERROR("Attempted to change expiration time to "
+          << next_expiration_time
+          << ", which precedes the previous expiration time of "
+          << *prev_expiration_time << ".");
+  }
+  *prev_expiration_time = next_expiration_time;
+}
+
+template <size_t MaxDerivPlusOne, bool StoreCoefs>
+const StoredInfo<MaxDerivPlusOne, StoreCoefs>& stored_info_from_upper_bound(
+    const double t, const std::vector<StoredInfo<MaxDerivPlusOne, StoreCoefs>>&
+                        all_stored_infos) noexcept {
+  // this function assumes that the times in stored_info_at_update_times is
+  // sorted, which is enforced by the update function of a piecewise polynomial.
+
+  if (all_stored_infos.empty()) {
+    ERROR(
+        "Vector of StoredInfos you are trying to access is empty. Was it "
+        "constructed properly?");
+  }
```

Comment:
So I did this, however, in one of my error tests now I'm running into an error in release mode where the ASSERT is basically just being ignored, regardless of whether the expression being evaluated is true or false.

Reply:
Oh wait, looking through the documentation, ASSERT doesn't do anything in release mode, it only checks in debug mode. In that case should I just get rid of it, or should I put an if statement with an ERROR to check that the vector isn't empty?

Reply:
for testing an ASSERT you should use ASSERTION_TEST instead of ERROR_TEST

Reply:
Yes if I'm using an assert, I should use an ASSERTION_TEST (however I tried this and got a `sys::abort("###No ASSERT tests in release mode ###")`. I guess the real question is how do I have a runtime assert statement that checks the size of my vector? The spectre ASSERT doesn't actually do anything in release mode, which is why I originally had the if statement then ERROR.

Reply:
whenever a condition should only be triggered by a coding error, then the condition should be checked in an ASSERT statement.  Developers should be compiling new code in debug mode and if the condition is violated it will be caught during development.  We use ASSERT instead of ERROR because we want to avoid testing the condition in release mode because repeatedly testing conditions is expensive and unnecessary.  ERROR is typically used to check dynamic conditions such as those use in root-finding or other numerical algorithms that fail and it is impossible to recover from.  So please change line 129 of the test to ASSERTION_ERROR and post (or point to on CI) the error.  Then we can try and figure out what is going wrong...

Reply:
There's some extra boilerplate required for ASSERT tests.  I think it's in the documentation somewhere (on testing?) or look for an example.

Reply:
yea just do this:

```
 SPECTRE_TEST_CASE(...) {
  ASSERTION_TEST();
#ifdef SPECTRE_DEBUG
  // testing code here...
  ERROR("Failed to trigger ASSERT in an assertion test");
#endif
}
```

Reply:
Checked this in release mode on my local branch and it worked 👍🏻